src/docs/emitter-redesign-proposal.md

```diff
@@ -0,0 +1,333 @@
+
+   Example:
+
+   ```swift
+   @frozen struct PersonWrapper {
+       let person: Person  // Person is a class
+       let id: Int
+   }
+   ```
+
+4. **Class** - Used for Swift classes, which are always reference types.
+
+   Example:
+
+   ```swift
+   class Animal {
+       var name: String
+   }
+   ```
+
+5. **Unknown** - Used for types that cannot be marshalled or are not supported.
+
+#### UNKNOWN Label Propagation Rules
+
+The `Unknown` label propagates according to these rules:
+
+1. If a type would be marshalled as something that requires layout (Struct, ClassWithBufferStruct) and any of its stored properties are `Unknown`, the type itself becomes `Unknown`.
+
+2. Generic types are currently marked as `Unknown` since we don't yet support them.
+
+Example of UNKNOWN propagation:
+
+```swift
+@frozen struct Container {
+    let value: GenericType<Int>  // GenericType is Unknown
+    let id: Int
+}
+// Container becomes Unknown because it has an Unknown field
+```
+
+The output of this step is an annotated graph where each node has a marshalling label. This information will be used in the subsequent pruning step.
+
+### Prune the graph and save data to the TypeDatabase
+
+The input of this step is the output of the last one – a collection of types with annotations how they should be marshalled.
+
+We should iterate over the collection and prune everything with `UNKNOWN` label. After pruning is completed a new TypeRecord for each remaining type should be created.
+
+TypeRecord should contain only information which is necessary for projecting "references" to a given type (e.g. when type occurs as a method argument or as a property / field type).
+
+Minimal viable type record:
+
+```csharp
+public record TypeRecord
+{
+    public required CSharpTypeName CSharpTypeName { get; init; }
+    public required SwiftTypeName SwiftTypeName { get; init; }
+    public required MarshallingLabel Label { get; init; }
+}
+```
+
+`MarshallingLabel` is the result of the previous step. It can be replaced by raw information about the type which allow to reconstruct the label later on (frozenness, requiresMemoryManagement). However the key thought behind this label is that things which share the same label should be marshalled the same - both declarations and references.
+
+## Type processing
+
+The purpose of this step is to process each type and its members, building a representation of how they should be marshalled.
+
+The input of this step is a pruned collection of TypeDecls and a TypeDatabase containing types from the current module.
+
+The simplest representation of a type is a `TypeRepresentation` which contains information about the type itself and its members.
+
+```csharp
+public record struct TypeRepresentation
+{
+    public required CSharpTypeName CSharpTypeName { get; init; }
+    public required MarshallingLabel Label { get; init; }
+    public required List<string> Methods { get; init; } 
```

Comment:
What is stored in List<string> Methods? Is it the source code of the methods which are emitted later?